### ISSUE TITLE
some virtualenv issues

### ISSUE THREAD

**Issue by aitchnyu** [DonJayamanne/pythonVSCode#148]
I use VS code 1.1.1, python extension 0.3.9 and Ubuntu 15.10. I got intellisense to work by settings.json with the following.

```
{
    "editor.rulers": [80,100],
    "python.pythonPath": "/home/jesvin/dev/ala/venv/bin/python",
    "python.autoComplete.extraPaths": [
        "/home/jesvin/dev/ala/venv/lib/python2.7",
        "/home/jesvin/dev/ala/venv/lib/python2.7/site-packages"
     ]
}

```
I then restarted VS de and then waited a while. Then intelliense and peek worked fine. I expected to see an "indexing python sources" while it was preparing indexes.

I then tried to use ${workspaceRoot}.  I set it to the following and restarted VS code and waited a while. Intellisense does not pick this up.

```
{
    "editor.rulers": [80,100],
    "python.pythonPath": "${workspaceRoot}/venv/bin/python",
    "python.autoComplete.extraPaths": [
        "${workspaceRoot}/venv/lib/python2.7",
        "${workspaceRoot}/venv/lib/python2.7/site-packages"
     ]
}
```

My issues are:

- Using virtualenv is mentioned in README.md in passing and not in the wiki.
-  Lack of feedback. I spent hours chasing down settings to see what works. I wanted to see some feedback like "please wait while we index your sources" or "the specified extraPaths /home/jesvin/dev/ala/venv/lib/python2.7/site-packages does not appear to contain Python packages".
- I did not get the ${workspaceRoot} version to work. I alternated between the two versions of settings.py and the one with absolute paths works consistently. I can help debug why.

These fixes could help:
- Wiki should have a complete example settings.json for Linux users of virtualenv. It should explicitly tell them that these settings isolate from packages in default interpreter.
- Wiki should tell users to restart VS code after changing settings and that the only feedback is that peek and intellisense starts to work after a while.
- The messages "please wait while we index your sources" and "the specified extraPaths /home/jesvin/dev/ala/venv/lib/python2.7/site-packages does not appear to contain Python packages"
- Have a python.virtualenv setting. This allows mistakes to be caught early. The messages "could not access the virtualenv path" and "the given path does not appear to be a virtualenv"
- Make the ${workspace} work. It helps with user-friendliness. For example, if I make a boilerplate, I dont want to have the instruction: "replace python.pythonPath and python.autoComplete.extraPaths to the absolute path of your newly created venv"

Sorry for rambling. Thanks for the extension and I look forward to using it to the fullest!

**Comment by DonJayamanne** [comment]
Hi, thank you very much for the feed back and all of the information.i will work on this asap.

**Comment by vanflymen** [comment]
It doesn't seem that `${workspaceRoot}` is works in `.vscode/settings.json`. Furthermore, I cannot get linting working. I've installed all the extra modules in the Virtualenv.

My `.vscode/settings.json` file:

```
{
    "editor.rulers": [90,110],
    "python.pythonPath": "C:\\Users\\Dan\\Documents\\GitHub\\my-proj\\.venv\\Scripts\\python.exe",
    "python.autoComplete.extraPaths": [
        "C:\\Users\\Dan\\Documents\\GitHub\\my-proj\\.venv\\Lib",
        "C:\\Users\\Dan\\Documents\\GitHub\\my-proj\\.venv\\Lib\\site-packages"
    ],
    "python.linting.pylintPath": "pylint --load-plugins pylint_django"
}
```

**Comment by vanflymen** [comment]
Nevermind. I figured out that you have to manually specify the paths for pep8, pylint etc. if you're using a Virtualenv... this is not a good idea.

It would be way better if we could use ${workspaceRoot} or specify the virtualenv and then $PATH should be amended accordingly.

**Comment by aitchnyu** [comment]
I guess that `${workspaceRoot}` is not substituted in the context of settings.json and PSVC trips on that value. That makes it an upstream issue. The virtualenv path should be sufficent to pick up the packages path as well as installed linters.

**Comment by DonJayamanne** [comment]
Thanks guys for the great feedback, i'm look into these issues. I haven't forgotten them.
Hopefully I'll have them wrapped up some time next week, including updating the WIKI.

**Comment by DonJayamanne** [comment]
Hi Everyone, I've updated the Wiki.
However as @vanflymen has pointed out, the linters will not work.
Also, ${workspaceRoot} is currently not supported. I will work on this soon.

Fyi - I'll also try to improve the configuration experience when dealing with Virtual Envs.

**Comment by DonJayamanne** [comment]
Hi everyone, you might want to try the following approach:

1. Clear all of the custom python paths in the settings.json (leave it to their defaults) and close VS Code
2. Open your terminal (command) window and activate the relevant python environment
```
(venv) ter@minal:~$ code project_folder/
```
3. Next, launch VS Code from that terminal (command window) session

Now, VS Code will inherit the environment variables (settings) configured for that particular Python Virtual Environment and everything should work as expected.
Please let me know how this goes.

Finally if this does work, then please let me know whether you prefer this approach or whether you would rather have some way of activating the environment from within VS Code & not have to configure the settings in settings.json.

**Comment by symbioticKid** [comment]
launched VScode as recommended from the command line after entering the virtualenvironment (This is windows)
took the simple example
```
value = 1 + 2
print(value)
```

F5 and debug console prints output 3. However, VScode prompts 'Error Canceled' Open launch.json.

```
{
            "name": "Python",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "program": "${file}",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput"
            ]
        },
```

**Comment by DonJayamanne** [comment]
1. Is there anything in the VS Code Debugger console?   
2. Please go into "Toggle Developer Tools" menu and let me know what errors are displayed in the Webkit in the console output?

**Comment by symbioticKid** [comment]
```
Canceled: Error: Canceled
    at Object.E [as create] (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:31:17647)
    at file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:104:23592
    at Object.p [as _notify] (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:16345)
    at Object.__dirname.undefined.F.enter (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:19673)
    at __dirname.undefined.t.Class.define._run (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:20994)
    at __dirname.undefined.t.Class.define._error (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:20511)
    at n (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:18216)
    at Object.p [as _notify] (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:16345)
    at Object.__dirname.undefined.F.enter (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:19673)
    at __dirname.undefined.t.Class.define._run (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:20994)
WARNING: Promise with no error callback:undefined
Error: Promise did not implement oncancel(…)
```

**Comment by DonJayamanne** [comment]
@symbioticKid , can you provide the following:
1. Version of python used
2. Does this happen if you aren't using any virtual environment? Standard debugging (launch VS Code from a non virtual environment)
3. Silly question, but are you able to run the program from the console window "Python xxx.py"

**Comment by symbioticKid** [comment]
1. 3.5.1
2. Works just fine outside of the virtual environment
3. [Ctrl + ~] inside of VScode. python simple.py -> get the output of 3. no issues

**Comment by DonJayamanne** [comment]
@symbioticKid , I have a windows 10 environment and it works as expected.
Sorry, but I will need some more information to get to the bottom of this:
1. When you say "works just find outside of the virtual environment"
Do you mean debugging works in VS Code, when not in a virtual environment?
2. Next how did you create the virtual environment? What did you use to create this?
3. Finally, have you tried setting up another virtual environment and testing this out?

Thanks for your help.

**Comment by symbioticKid** [comment]
Sorry for the ambiguity.
1. Correct, when I point VScode to a folder with no virtual environment, I'm able to hit F5 and step through each line of the code. virtual environment hasn't been created for that folder and launching code from the regular command prompt and not (venv) C:/xxx

2. Steps below
C:\xxx\L1dir> virtualenv L2dir
C:\xxx\L1dir\L2dir> Scripts\activate
(L2dir) C:\xxx\L1dir> code L2dir

3. Tried in two separate virtualenv and get the above mentioned error.

Hopefully this helps.

**Comment by DonJayamanne** [comment]
@symbioticKid , please could you ensure all instances of VS Code are closed before launching it from the command prompt.

**Comment by symbioticKid** [comment]
Just tried it again - all instances of vscode were closed and the new one was launched from the context of the virtualenv.

launch.json
```
"name": "Python",
            "type": "python",
            "pythonPath": "python",
            "request": "launch",
            "stopOnEntry": true,
            "program": "${file}",
            "externalConsole": false,
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput"
            ]

```

simple debug program:
```
import sys
print(sys.prefix)
print(sys.executable)

```

output being printed in debug-console:
```
C:\Users\xxxxx\XXXXX\venv
C:\Users\xxxxx\XXXXX\venv\Scripts\python.exe

```

And the error:
```
/C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:87 Canceled: Error: Canceled
    at Object.E [as create] (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:31:17647)
    at file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:104:23592
    at Object.p [as _notify] (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:16345)
    at Object.__dirname.undefined.F.enter (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:19673)
    at __dirname.undefined.t.Class.define._run (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:20994)
    at __dirname.undefined.t.Class.define._error (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:20511)
    at n (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:18216)
    at Object.p [as _notify] (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:16345)
    at Object.__dirname.undefined.F.enter (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:19673)
    at __dirname.undefined.t.Class.define._run (file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:23:20994)e.doShow @ /C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:87
/C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:32 WARNING: Promise with no error callback:undefined
/C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:32 Error: Promise did not implement oncancel(…)
```

At this point, i'd say the launching the debugger in the virtualenv seems to be the issue in my case. When I execute the same sample code without the virtualenv, i'm able to go into the debugger.

**Comment by DonJayamanne** [comment]
@symbioticKid , I have tried creating the virtual environment using your instructions.
I have pushed through a new update. please install that and let me know how it goes.

1. Also I'm a little confused. At what stage does the above error get displayed.
You've indicated that the two print statements are working (printing of the path and the executable path)!
So, at what point do you get the "Cancelled" error message?

2. Could you try the following code in the virtual environment with the new update:
import sys
print(sys.prefix)
print(sys.executable)
print(sys.version)
value = 1 + 2
print(value)

**Comment by symbioticKid** [comment]
1. Right after starting the debug (F5). With "stopOnEntry": true and pythonVSCode version 0.3.10, I'd not be able to do F11 or F10 (step into or step over), only the output on the debug console. So the print statements print to the debug console and the error: canceled message shows up at the top of the screen.
However, when i'm launching the debug (F5) outside of the virtualenv, i'm able to step into (F11) and stepover (F10) the piece of code and no error: canceled message pops from the top of the screen.

With version 0.3.11, the error message doesn't show up anymore when launched from within the virtualenv, however F11 and F10 don't seem to be working. Outside of virtualenv, both F11 and F10 work.

Steps below: 
<img width="671" alt="01 cmdline" src="https://cloud.githubusercontent.com/assets/613951/16185530/5926f89a-3679-11e6-9e4b-49aad08bd5f4.PNG">

<img width="1254" alt="launch-json" src="https://cloud.githubusercontent.com/assets/613951/16185606/e8a040d0-3679-11e6-8550-3e9a4812e454.png">

<img width="1252" alt="debug" src="https://cloud.githubusercontent.com/assets/613951/16185610/ee60cc24-3679-11e6-881a-ff17573537d3.PNG">
With 0.3.11, stepinto and step over don't seem to be working, although the error:canceled message has gone

**Comment by DonJayamanne** [comment]
@vanflymen 
Paths relative to the workspace root are currently supported in the "python.autoComplete.extraPaths"
However I will be adding (today) support for ${workspaceRoot} in the paths. 

Finally, I'm also making a lot of changes around how paths are resolved when dealing with virtual environments. 
The goal is for the users to have to configure (define) the python path in one place (not for all of pylint, auto complete, etc).

**Comment by DonJayamanne** [comment]
@aitchnyu , I'm working on improving the support for virtual environments.  
My plan is to get the end user to configure just the python.pythonPath setting in setting.json.   

I am working on a troubleshooting section in the wiki and will update the wiki (home page) with more (simple) instructions on configuring the IDE for virtual environments.

**Comment by DonJayamanne** [comment]
@aitchnyu , @vanflymen , @symbioticKid I have uploaded a new version of the extension.
Hopefully this will fix most of the virtual env issues.
All you now need to do is:
* Configure the "python.pythonPath" setting in settings.json
* All other paths such as linters, formatters can remain at their default. I.e. if you have provided custom paths for them, please remove them (not necessary).

If you have any issues, please report them here, i'd like to understand them and hopefully it would be a configuration issue.

Once again, thanks a lot for your feed back in helping me get this sorted out.

**Comment by symbioticKid** [comment]
Installed the 0.3.13 update. Still seeing same behavior as before - step-in and step-over don't work.

Activated the virtual environment via cmd prompt, but didn't launch vscode from there. Open vscode seperately and added the following to workspace settings.json file.

`
    "python.pythonPath": "C:/venv/Scripts/python.exe"
`

Just the above change didn't seem to get picked up by launch.json. So had to add the change there as well.

Ran the following code:
```
import sys
print(sys.prefix)
print(sys.executable)
print(sys.version)
```

1. With changes in both launch.json and settings.json, still seeing the same behavior where step-in and step-over don't work
2. By not launching vscode from the virtual environment cmd prompt, the in terminal command prompt (Ctrl + ~) doesn't point to the virtual environment but the default folder and python installation.

**Comment by DonJayamanne** [comment]
>Just the above change didn't seem to get picked up by launch.json.    

That's correct, you will still need to define this in launch.json.

> _1. still seeing the same behavior where step-in, step-over and breakpoints don't work_    

Highly unusual. I'm at a loss at this stage.
Please could you let me know the version of VS Code you have installed? 
And the only other suggestion I have is, please move the code out of the virtual environment directory.  
Right now the virtual enviroment is "c:/venv"  
Can I suggest you try moving the code into c:/test or similar (i.e. another directory).  

> _2.  So no (venv) C:<etc>, but the default C:<etc>_

Yes, this is correct. The terminal has nothing to do with the python path defined.  
If you want the terminal to inherit your virtual enviroment then you must:
 + Close all instances of VS Code
 + Go into a terminal window, activate your virtual env
 + Next launch VS Code form within the terminal session

The terminal widow has nothing to do with the path to the python executable. That's just a path to an executable, and doesn't define the virtual environment. Its the bash script in bin/activate that would activate/define the virtual environment.

**Comment by symbioticKid** [comment]
<img width="191" alt="vs version" src="https://cloud.githubusercontent.com/assets/613951/16308864/7b07bb9e-391b-11e6-9db9-9c993ed77fb1.PNG">

**Comment by symbioticKid** [comment]
Moved to debug script from venv to a new directory C:\test. Created a new virtual environment there and launched code from the command prompt. Still see the same failing behavior with no step-in, step-over or stopping at breakpoints :(

Not sure if it's just me seeing this, but happy to provide any additional logs you'd want

**Comment by aitchnyu** [comment]
Let me notify you about a stackoverflow question I asked about the same http://stackoverflow.com/q/37642045/604511 which received a bit of traffic already. Please add the correct answer when its time. I will check out your improvements soon, its a hobby project.

**Comment by DonJayamanne** [comment]
@aitchnyu , a new build has been released and this should be resolved.

**Comment by DonJayamanne** [comment]
I will be closing this issue, as this has been resolved. If the problem still persists, please feel free to open it.

**Comment by snff** [comment]
Hello Don, 
Great plugin, but it seems like there is still an issue with using virtual envs. I just downloaded the extension and tried various different ways from the documentation and not much luck. It will be great to have you post some sample launch.json and settings.json with correct settings that can work.

**Comment by uralbash** [comment]
@snff 

``.vscode/settings.json``
```python
{
    "editor.fontSize": 18,

    //Python
    "python.pythonPath": "${workspaceRoot}/myenv/bin/python",

    // Whether to lint Python files using pylint.
    "python.linting.pylintEnabled": true,

    // Whether to lint Python files using pep8
    "python.linting.pep8Enabled": true,

    // Whether to lint Python files using flake8
    "python.linting.flake8Enabled": true
}
```

``.vscode/launch.json``
```python
{
    "version": "0.2.0",
    "configurations": [{
        "name": "Pyramid",
        "type": "python",
        "request": "launch",
        "stopOnEntry": true,
        "pythonPath": "${workspaceRoot}/myenv/bin/python",
        "program": "${workspaceRoot}/myenv/bin/pserve",
        "args": ["${workspaceRoot}/development.ini"],
        "debugOptions": [
            "WaitOnNormalExit",
            "RedirectOutput"
        ]
    }]
}
```

**Comment by snff** [comment]
Thanks did you get inline debugging working with this config. I have a flask config. Every time I try to debug it cannot see the Flask module, so it fails on the import. I know it is there because when I run it directly in a virtual env shell it works fine.

**Comment by uralbash** [comment]
To make debug work, you need to set ``pythonPath`` option in ``.vscode/launch.json``. For lint and gotodefinition enable, you also need to set ``python.pythonPath`` in ``.vscode/settings.json``. It works for me. Don't forget restart editor.

**Comment by snff** [comment]
Great, that was the key.  Appreciate your quick response. I am past the initial hick up I had. I still can't get the break points to function. Anything special for that ?

**Comment by villancikos** [comment]
If anyone comes with the same issue and wants a TL;DR:
Just open your Project Folder from Terminal with the virtual environment running 
`(myvirtualenv) user$ code .`

**Comment by u84six** [comment]
I've been trying to get this to work with vscode and virtualenv on mac. When I launch from venv directory after activating, it seems to work, but how do I know if it's using the virtualenv python and not global? When I hardcode the path to my venv python location in settings.json, I see a crash when debugging.

**Comment by DonJayamanne** [comment]
I would not recommend activating and launching. Laugh vscode and modify settings.json to point to your Python interpreter.

**Comment by haarburger** [comment]
I've got a very similar problem with the latest version of pythonVSCode and Python 3.6 (Anaconda):

A custom package is added to my python path by `conda develop /path/to/my_package`. This generally works well but in VSCode, autocomplete does not work for the package when using absolute imports:

```python
from my_package import some_function
```
I started VSCode from bash with the correct conda environment activated. Strangely, if I execute `python -c "import sys; print('\n'.join(sys.path))"` from the integrated terminal, `/path/to/my/package` is not included, even though it is when executed from my system terminal.
